File: 03-poo/07-sobrescritura-y-polimorfismo/java/soluciones/RetoNotificaciones.java
public class RetoNotificaciones {
    interface Notificable { void enviar(String mensaje); }
    static class Email implements Notificable { public void enviar(String mensaje){ System.out.println("Email: " + mensaje); } }
    static class Mensajeria implements Notificable { public void enviar(String mensaje){ System.out.println("Chat: " + mensaje); } }
    public static void main(String[] args) {
        Notificable[] canales = {new Email(), new Mensajeria()};
        for (Notificable canal : canales) canal.enviar("Curso disponible");
    }
}

File: 03-poo/07-sobrescritura-y-polimorfismo/scala/soluciones/RetoNotificaciones.scala
trait Notificable:
  def enviar(mensaje: String): Unit
class EmailReto extends Notificable:
  def enviar(mensaje: String): Unit = println(s"Email: $mensaje")
class ChatReto extends Notificable:
  def enviar(mensaje: String): Unit = println(s"Chat: $mensaje")

@main def retoNotificaciones(): Unit = List[Notificable](new EmailReto,new ChatReto).foreach(_.enviar("Curso disponible"))
